opts/filters.go

```diff
@@ -0,0 +1,98 @@
+package dhtopts
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+// PublicQueryFilter returns true if the peer is suspected of being publicly accessible
+func PublicQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range ai.Addrs {
+		if isRelayAddr(a) {
+			return false
+		}
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+	return hasPublicAddr
+}
+
+// PublicRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a public network
+func PublicRoutingTableFilter(conns []network.Conn) bool {
+	for _, c := range conns {
+		addr := c.RemoteMultiaddr()
+		if !isRelayAddr(addr) && manet.IsPublicAddr(addr) {
+			return true
```

Comment:
This may be a bit more aggressive than necessary since AFAIK if two peers are publicly dialable via port forwarding and are also on the same local network they likely will be connected to each other over private multiaddrs instead of public ones. This means we are potentially missing peers from our routing tables that could be there.

In the future we could add more expensive tests like pinging any non-relay public addresses we have for the peer in the peerstore (requires that the function have access to the peerstore), to see if the peer we're locally connected to is also publicly dialable.

If desired there could also be a simple check to see if (as far as we know) the peer and us share the same public network address (e.g. are behind the same NAT).

Reply:
Can we just check "do we have a public addr for this peer in the peerstore"?

Reply:
@aschmahmann 

Would I be correct in saying that the problem we want to solve is :

"Even for a public DHT, we should add a peer to the RT if that peer is on the same network as ours" ?

In

Reply:
@aarshkshah1992 The problem we're trying to solve in this function (PublicRoutingTableFilter) is:

How can we help ourselves ensure (at minimal cost) that even if peers are not properly removing themselves from the public DHT (i.e. switching to client mode if they are undialable) that we do not add non-publicly dialable peers to our routing table?